src/py21cmfast/src/SpinTemperatureBox.c

```diff
@@ -1949,29 +2084,60 @@ LOG_SUPER_DEBUG("looping over box...");
 
                             // next, Compton heating
                             //                dcomp_dzp = dT_comp(zp, T, x_e);
-                            dcomp_dzp = dcomp_dzp_prefactor*(x_e/(1.0+x_e+f_He))*( Trad_fast - T );
+                            dcomp_dzp = dcomp_dzp_prefactor*(x_e/(1.0+x_e))*( Trad_fast - T );//removed f_He
+
 
                             // lastly, X-ray heating
                             dxheat_dzp = dxheat_dt_box[box_ct] * dt_dzp * 2.0 / 3.0 / k_B / (1.0+x_e);
                             if (flag_options->USE_MINI_HALOS){
                                 dxheat_dzp_MINI = dxheat_dt_box_MINI[box_ct] * dt_dzp * 2.0 / 3.0 / k_B / (1.0+x_e);
                             }
 
+                            //CMB heating rate
+							dCMBheat_dzp = 0.;
+                            if (flag_options->USE_CMB_HEATING) {
+								//Meiksin et al. 2021
+								eps_CMB = (3./4.) * (T_cmb*(1.+zp)/T21) * A10_HYPERFINE * f_H * (hplank*hplank/Lambda_21/Lambda_21/m_p) * (1.+2.*T/T21);
+								dCMBheat_dzp = 	-eps_CMB * (2./3./k_B/(1.+x_e))/hubble(zp)/(1.+zp);
+                            }
+                            //Ly-alpha heating rate
+							eps_Lya_cont = 0.;
+							eps_Lya_inj = 0.;
+							eps_Lya_cont_MINI = 0.;
+							eps_Lya_inj_MINI = 0.;
+                            if (flag_options->USE_Lya_HEATING) {
+                                E_continuum = Energy_Lya_heating(T, previous_spin_temp->Ts_box[box_ct], taugp(zp,curr_delNL0*growth_factor_zp,x_e), 2);
+                                E_injected = Energy_Lya_heating(T, previous_spin_temp->Ts_box[box_ct], taugp(zp,curr_delNL0*growth_factor_zp,x_e), 3);
+                                Ndot_alpha_cont = (4.*PI*Ly_alpha_HZ) / (N_b0*pow(1.+zp,3.)*(1.+curr_delNL0*growth_factor_zp))/(1.+zp)/C * dstarlya_cont_dt_box[box_ct];
+                                Ndot_alpha_inj = (4.*PI*Ly_alpha_HZ) / (N_b0*pow(1.+zp,3.)*(1.+curr_delNL0*growth_factor_zp))/(1.+zp)/C * dstarlya_inj_dt_box[box_ct];
+                                eps_Lya_cont = - Ndot_alpha_cont * E_continuum * (2. / 3. /k_B/ (1.+x_e));
```

Comment:
again, it depends on what Energy_Lya_heating gives, do we need to consider f_He here?

Reply:
No, we should not. The x_e is defined as n_e/(n_H+n_He) in 21cmFAST. If you take f_He in the denominator, you'll have excess He count. The denominator should be n_tot = n_e + n_H + n_He. So n_tot = (n_H + n_He)(1+x_e). If you define the fractions in terms of n_H, then you'll have f_He, and in that case your n_tot = n_H (1+x_e+f_He). For example, see  T. Venumadhav + Rev. D 98, 103513 (2018) this paper.